Spigot-API-Patches/0305-per-viewer-chat-rendering.patch

```diff
@@ -0,0 +1,356 @@
++++ b/src/main/java/io/papermc/paper/event/player/AsyncChatEvent.java
+@@ -3,6 +3,8 @@ package io.papermc.paper.event.player;
+ import io.papermc.paper.chat.ChatComposer;
+ import io.papermc.paper.chat.ChatFormatter;
+ import java.util.Set;
++import io.papermc.paper.chat.ChatRenderer;
++import net.kyori.adventure.audience.Audience;
+ import net.kyori.adventure.text.Component;
+ import org.bukkit.entity.Player;
+ import org.bukkit.event.HandlerList;
+@@ -14,12 +16,20 @@ import org.jetbrains.annotations.NotNull;
+ public final class AsyncChatEvent extends AbstractChatEvent {
+     private static final HandlerList HANDLERS = new HandlerList();
+ 
++    public AsyncChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull Set<Audience> viewers, final @NotNull ChatRenderer renderer, final @NotNull Component message) {
++        super(async, player, recipients, viewers, renderer, message);
++    }
++
++    /**
++     * @deprecated use {@link #AsyncChatEvent(boolean, Player, Set, Set, ChatRenderer, Component)}
++     */
++    @Deprecated
+     public AsyncChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull ChatComposer composer, final @NotNull Component message) {
+         super(async, player, recipients, composer, message);
+     }
+ 
+     /**
+-     * @deprecated use {@link #AsyncChatEvent(boolean, Player, Set, ChatComposer, Component)}
++     * @deprecated use {@link #AsyncChatEvent(boolean, Player, Set, Set, ChatRenderer, Component)}
+      */
+     @Deprecated
+     public AsyncChatEvent(final boolean async, final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull ChatFormatter formatter, final @NotNull Component message) {
+diff --git a/src/main/java/io/papermc/paper/event/player/ChatEvent.java b/src/main/java/io/papermc/paper/event/player/ChatEvent.java
+index 13c5df5fb8ce1d0203d99e88dd691019146a8f52..e2e4cd5efad14d24197c9a39f2d2664baeafc782 100644
+--- a/src/main/java/io/papermc/paper/event/player/ChatEvent.java
++++ b/src/main/java/io/papermc/paper/event/player/ChatEvent.java
+@@ -3,6 +3,8 @@ package io.papermc.paper.event.player;
+ import io.papermc.paper.chat.ChatComposer;
+ import io.papermc.paper.chat.ChatFormatter;
+ import java.util.Set;
++import io.papermc.paper.chat.ChatRenderer;
++import net.kyori.adventure.audience.Audience;
+ import net.kyori.adventure.text.Component;
+ import org.bukkit.Warning;
+ import org.bukkit.entity.Player;
+@@ -19,12 +21,20 @@ import org.jetbrains.annotations.NotNull;
+ public final class ChatEvent extends AbstractChatEvent {
+     private static final HandlerList HANDLERS = new HandlerList();
+ 
++    public ChatEvent(final @NotNull Player player, final @NotNull Set<Player> recipients, final @NotNull Set<Audience> viewers, final @NotNull ChatRenderer renderer, final @NotNull Component message) {
```

Comment:
Do we want this constructor to take `Set<Player> recipients`, or just `Set<Audience> viewers`?

Reply:
Sadly in order to avoid eagerly initializing these sets (in impl they are lazy sets which do not populate until accessed) we need to pass both in.